KeyError: 'publish' (lib\site-packages\pybuilder\execution.py:292)

I faced the following error after upgrading the pybuilder from 0.11 to 0.12 (environment 'system': CPython version 3.6.8.final.0 on win32).
<pre>
PyBuilder version 0.12.6
Build started at 2020-07-20 22:30:50
------------------------------------------------------------
[INFO]  Installing or updating plugin "pypi:my_pybuilder_plugin, module name 'my_pybuilder_plugin'"
[INFO]  Processing plugin packages 'my_pybuilder_plugin' to be installed with {}
------------------------------------------------------------
BUILD FAILED - KeyError: 'publish' (lib\site-packages\pybuilder\execution.py:292)
------------------------------------------------------------
Build finished at 2020-07-20 22:31:09
Build took 19 seconds (19333 ms)
</pre>

I'm not sure that I fully understand the method [pybuilder.execution.ExecutionManager.register_late_task_dependencies](https://github.com/pybuilder/pybuilder/blob/master/src/main/python/pybuilder/execution.py#L283-L294) but it seems to me there is a error parameter in the debug message. Please look at [the line 292](https://github.com/pybuilder/pybuilder/blob/master/src/main/python/pybuilder/execution.py#L292)) or check it in this pull request #738 .

I suppose it is necessary to use `self._dependencies_pending_tasks` instead of `dependencies` in this case. At least my project was built successfully after this fix.

I'm sorry, Arcadiy, but I have no possibility to share the code of this plugin.

The build log is here: 
<pre>
(venv_win_3.6) D:\src\EAS\aws\eas-redirector-lambda>pyb -v -X
[DEBUG] Registered Python environment 'system': CPython version 3.6.8.final.0 on win32 in ['d:\\src\\eas\\aws\\eas-redirector-lambda\\venv_win_3.6\\scripts\\python.exe']
PyBuilder version 0.12.6
Build started at 2020-07-21 10:25:50
------------------------------------------------------------
[DEBUG] Loading project module from .\build.py
[DEBUG] Setting up plugins VEnv at 'd:\src\eas\aws\eas-redirector-lambda\.pybuilder\plugins\cpython-3.6.8.final.0'
[DEBUG] Registered Python environment 'pybuilder': CPython version 3.6.8.final.0 on win32 in ['d:\\src\\eas\\aws\\eas-redirector-lambda\\.pybuilder\\plugins\\cpython-3.6.8.final.0\\script
s\\python.exe']
[DEBUG] Patched __import__ system to defer plugin loading
[INFO]  Installing or updating plugin "pypi:pybuilder_defaults_plugin, module name 'pybuilder_defaults_plugin'"
[DEBUG] Inspecting package pybuilder_defaults_plugin
[DEBUG] Package pybuilder_defaults_plugin is not installed and will be installed
[INFO]  Processing plugin packages 'pybuilder_defaults_plugin' to be installed with {}
[DEBUG] Executing command: 'd:\\src\\eas\\aws\\eas-redirector-lambda\\.pybuilder\\plugins\\cpython-3.6.8.final.0\\scripts\\python.exe' '-m' 'pip.__main__' 'install' 'pybuilder_def
aults_plugin'
[DEBUG] Loading plugin 'pypi:pybuilder_defaults_plugin'
[DEBUG] Trying to load third party plugin 'pypi:pybuilder_defaults_plugin', module 'pybuilder_defaults_plugin'
[DEBUG] Found third party plugin 'pypi:pybuilder_defaults_plugin', module 'pybuilder_defaults_plugin'
[DEBUG] Loading plugin 'filter_resources'
[DEBUG] Trying to load direct plugin 'filter_resources', module 'pybuilder.plugins.filter_resources_plugin'
[DEBUG] Found direct plugin 'filter_resources', module 'pybuilder.plugins.filter_resources_plugin'
[DEBUG] Loading plugin 'copy_resources'
[DEBUG] Trying to load direct plugin 'copy_resources', module 'pybuilder.plugins.copy_resources_plugin'
[DEBUG] Found direct plugin 'copy_resources', module 'pybuilder.plugins.copy_resources_plugin'
[DEBUG] Loading plugin 'python.core'
[DEBUG] Trying to load direct plugin 'python.core', module 'pybuilder.plugins.python.core_plugin'
[DEBUG] Found direct plugin 'python.core', module 'pybuilder.plugins.python.core_plugin'
[DEBUG] Loading plugin 'python.install_dependencies'
[DEBUG] Trying to load direct plugin 'python.install_dependencies', module 'pybuilder.plugins.python.install_dependencies_plugin'
[DEBUG] Found direct plugin 'python.install_dependencies', module 'pybuilder.plugins.python.install_dependencies_plugin'
[DEBUG] Loading plugin 'python.distutils'
[DEBUG] Trying to load direct plugin 'python.distutils', module 'pybuilder.plugins.python.distutils_plugin'
[DEBUG] Found direct plugin 'python.distutils', module 'pybuilder.plugins.python.distutils_plugin'
[DEBUG] Loading plugin 'python.unittest'
[DEBUG] Trying to load direct plugin 'python.unittest', module 'pybuilder.plugins.python.unittest_plugin'
[DEBUG] Found direct plugin 'python.unittest', module 'pybuilder.plugins.python.unittest_plugin'
[DEBUG] Loading plugin 'python.integrationtest'
[DEBUG] Trying to load direct plugin 'python.integrationtest', module 'pybuilder.plugins.python.integrationtest_plugin'
[DEBUG] Found direct plugin 'python.integrationtest', module 'pybuilder.plugins.python.integrationtest_plugin'
[DEBUG] Loading plugin 'python.coverage'
[DEBUG] Trying to load direct plugin 'python.coverage', module 'pybuilder.plugins.python.coverage_plugin'
[DEBUG] Found direct plugin 'python.coverage', module 'pybuilder.plugins.python.coverage_plugin'
[DEBUG] Loading plugin 'python.flake8'
[DEBUG] Trying to load direct plugin 'python.flake8', module 'pybuilder.plugins.python.flake8_plugin'
[DEBUG] Found direct plugin 'python.flake8', module 'pybuilder.plugins.python.flake8_plugin'
[DEBUG] Loading plugin 'core'
[DEBUG] Trying to load direct plugin 'core', module 'pybuilder.plugins.core_plugin'
[DEBUG] Found direct plugin 'core', module 'pybuilder.plugins.core_plugin'
[DEBUG] Loading plugin 'analysis'
[DEBUG] Trying to load direct plugin 'analysis', module 'pybuilder.plugins.analysis_plugin'
[DEBUG] Found direct plugin 'analysis', module 'pybuilder.plugins.analysis_plugin'
[DEBUG] Found task 'prepare' with dependencies []
[DEBUG] Registering task 'prepare'
[DEBUG] Registering initializer 'init_python_directories'
[DEBUG] Found task 'package' with dependencies []
[DEBUG] Registering task 'package'
[DEBUG] Found task 'prepare' with dependencies []
[DEBUG] Registering task 'prepare'
[DEBUG] Registering initializer 'initialize_install_dependencies_plugin'
[DEBUG] Found task 'install_build_dependencies' with dependencies [prepare]
[DEBUG] Registering task 'install_build_dependencies'
[DEBUG] Found task 'install_dependencies' with dependencies [prepare]
[DEBUG] Registering task 'install_dependencies'
[DEBUG] Found task 'install_runtime_dependencies' with dependencies [prepare]
[DEBUG] Registering task 'install_runtime_dependencies'
[DEBUG] Found task 'list_dependencies' with dependencies []
[DEBUG] Registering task 'list_dependencies'
[DEBUG] Found action build_binary_distribution
[DEBUG] Registering action 'build_binary_distribution'
[DEBUG] Registering initializer 'initialize_distutils_plugin'
[DEBUG] Found task 'install' with dependencies []
[DEBUG] Registering task 'install'
[DEBUG] Found action set_description
[DEBUG] Registering action 'set_description'
[DEBUG] Found task 'upload' with dependencies []
[DEBUG] Registering task 'upload'
[DEBUG] Found action write_manifest_file
[DEBUG] Registering action 'write_manifest_file'
[DEBUG] Found action write_setup_script
[DEBUG] Registering action 'write_setup_script'
[DEBUG] Found action coverage_init
[DEBUG] Registering action 'coverage_init'
[DEBUG] Registering initializer 'init_test_source_directory'
[DEBUG] Found task 'run_unit_tests' with dependencies []
[DEBUG] Registering task 'run_unit_tests'
[DEBUG] Found action coverage_init
[DEBUG] Registering action 'coverage_init'
[DEBUG] Registering initializer 'initialize_integrationtest_plugin'
[DEBUG] Found task 'run_integration_tests' with dependencies []
[DEBUG] Registering task 'run_integration_tests'
[DEBUG] Found task 'analyze' with dependencies [run_unit_tests]
[DEBUG] Registering task 'analyze'
[DEBUG] Found task 'coverage' with dependencies [verify]
[DEBUG] Registering task 'coverage'
[DEBUG] Registering initializer 'init_coverage_properties'
[DEBUG] Found task 'prepare' with dependencies []
[DEBUG] Registering task 'prepare'
[DEBUG] Registering late dependency of task 'publish' on [coverage(optional)]
[DEBUG] Found task 'analyze' with dependencies [prepare]
[DEBUG] Registering task 'analyze'
[DEBUG] Found action assert_flake8_is_executable
[DEBUG] Registering action 'assert_flake8_is_executable'
[DEBUG] Registering initializer 'initialize_flake8_plugin'
[DEBUG] Found task 'analyze' with dependencies [prepare]
[DEBUG] Registering task 'analyze'
[DEBUG] Found task 'build_lambda_package' with dependencies [clean, install_dependencies, verify, publish]
[DEBUG] Registering task 'build_lambda_package'
[DEBUG] Found action configure_pylint
[DEBUG] Registering action 'configure_pylint'
[DEBUG] Found action git_release
[DEBUG] Registering action 'git_release'
[DEBUG] Registering initializer 'initialize_default_configuration'
[DEBUG] Found task 'prepare_release' with dependencies []
[DEBUG] Registering task 'prepare_release'
[DEBUG] Found task 'release' with dependencies [prepare_release, clean, install_build_dependencies, install_runtime_dependencies, verify, publish]
[DEBUG] Registering task 'release'
[DEBUG] Found task 'release_lambda' with dependencies [prepare_release, clean, install_build_dependencies, install_runtime_dependencies, verify, publish, build_lambda_package]
[DEBUG] Registering task 'release_lambda'
[DEBUG] Found task 'verify' with dependencies [run_unit_tests, run_integration_tests, analyze]
[DEBUG] Registering task 'verify'
[DEBUG] Found task 'clean' with dependencies []
[DEBUG] Registering task 'clean'
[DEBUG] Found task 'compile_sources' with dependencies [prepare]
[DEBUG] Registering task 'compile_sources'
[DEBUG] Registering initializer 'init'
[DEBUG] Found task 'install' with dependencies [package, publish(optional)]
[DEBUG] Registering task 'install'
[DEBUG] Found task 'package' with dependencies [compile_sources, run_unit_tests(optional)]
[DEBUG] Registering task 'package'
[DEBUG] Found task 'prepare' with dependencies []
[DEBUG] Registering task 'prepare'
[DEBUG] Found task 'print_module_path' with dependencies []
[DEBUG] Registering task 'print_module_path'
[DEBUG] Found task 'print_scripts_path' with dependencies []
[DEBUG] Registering task 'print_scripts_path'
[DEBUG] Found task 'publish' with dependencies [package, verify(optional)]
[DEBUG] Registering task 'publish'
[DEBUG] Found task 'run_integration_tests' with dependencies [package]
[DEBUG] Registering task 'run_integration_tests'
[DEBUG] Found task 'run_unit_tests' with dependencies [compile_sources]
[DEBUG] Registering task 'run_unit_tests'
[DEBUG] Found task 'verify' with dependencies [run_integration_tests(optional)]
[DEBUG] Registering task 'verify'
Traceback (most recent call last):
  File "d:\src\eas\aws\eas-redirector-lambda\venv_win_3.6\lib\site-packages\pybuilder\cli.py", line 460, in main
    offline=options.offline
  File "d:\src\eas\aws\eas-redirector-lambda\venv_win_3.6\lib\site-packages\pybuilder\reactor.py", line 210, in prepare_build
    self._collect_project_annotations()
  File "d:\src\eas\aws\eas-redirector-lambda\venv_win_3.6\lib\site-packages\pybuilder\reactor.py", line 560, in _collect_project_annotations
    self.collect_project_annotations(deferred_plugin[1])
  File "d:\src\eas\aws\eas-redirector-lambda\venv_win_3.6\lib\site-packages\pybuilder\reactor.py", line 394, in collect_project_annotations
    self.execution_manager.register_late_task_dependencies(injected_task_dependencies)
  File "d:\src\eas\aws\eas-redirector-lambda\venv_win_3.6\lib\site-packages\pybuilder\execution.py", line 292, in register_late_task_dependencies
    self.logger.debug("Resolved late dependency of task '%s' on %s", name, dependencies[name])
KeyError: 'publish'
------------------------------------------------------------
BUILD FAILED - KeyError: 'publish' (lib\site-packages\pybuilder\execution.py:292)
------------------------------------------------------------
Build finished at 2020-07-21 10:26:09
Build took 19 seconds (19470 ms)
</pre>

--- a/src/integrationtest/python/issue_737_tests.py
+++ b/src/integrationtest/python/issue_737_tests.py
@@ -1,0 +1,75 @@
+#   -*- coding: utf-8 -*-
+#
+#   This file is part of PyBuilder
+#
+#   Copyright 2011-2020 PyBuilder Team
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+
+import unittest
+
+from itest_support import IntegrationTestSupport
+
+
+class Test(IntegrationTestSupport):
+    def test(self):
+        self.write_build_file("""
+from pybuilder.core import use_plugin, init
+import sys
+from os.path import join as jp, normcase as nc, dirname
+
+build_sources = nc(jp(dirname(__file__), "src/main/python"))
+if build_sources not in sys.path:
+    sys.path.insert(0, build_sources)
+
+use_plugin("issue_737_plugin_2")
+use_plugin("issue_737_plugin_1")
+
+name = "pybuilder-defaults-plugin"
+version = "1.1.1"
+summary = "PyBuilder plugin to provide common configuration for python projects"
+default_task = ["task2"]
+
+@init
+def init (project):
+    pass
+""")
+        self.create_directory("src/main/python")
+        self.write_file("src/main/python/issue_737_plugin_1.py", r"""
+from pybuilder.core import init, task, depends
+
+@task
+def task1():
+    pass
+
+@task
+@depends("task1")
+def task2():
+    pass
+""")
+        self.write_file("src/main/python/issue_737_plugin_2.py", r"""
+from pybuilder.core import init, use_plugin, task, depends, dependents, optional
+
+@task
+@depends("task1")
+@dependents(optional("task2"))
+def taskX():
+    pass
+""")
+        reactor = self.prepare_reactor()
+        reactor.build()
+
+
+if __name__ == "__main__":
+    unittest.main()

--- a/src/main/python/pybuilder/execution.py
+++ b/src/main/python/pybuilder/execution.py
@@ -122,7 +122,6 @@ class Action(Executable):
 class TaskDependency(object):
     def __init__(self, mixed, optional=False):
         self._name = as_task_name(mixed)
-        self._task = mixed if hasattr(mixed, "name") else None
         self._optional = optional
 
     def __repr__(self):
@@ -137,12 +136,11 @@ class TaskDependency(object):
         return self._name
 
     @property
-    def task(self):
-        return self._task
-
-    @property
     def optional(self):
         return self._optional
+
+    def required(self):
+        self._optional = False
 
 
 class Task(object):
@@ -282,15 +280,17 @@ class ExecutionManager(object):
 
     def register_late_task_dependencies(self, dependencies):
         for name in dependencies:
-            self.logger.debug("Registering late dependency of task '%s' on %s", name, dependencies[name])
+            dependency_tasks = dependencies[name]
+            self.logger.debug("Registering late dependency of task '%s' on %s", name, dependency_tasks)
             if name not in self._dependencies_pending_tasks:
                 self._dependencies_pending_tasks[name] = []
-            self._dependencies_pending_tasks[name].extend(dependencies[name])
+            self._dependencies_pending_tasks[name].extend(dependency_tasks)
 
         for name in list(self._dependencies_pending_tasks.keys()):
             if self.has_task(name):
-                self.logger.debug("Resolved late dependency of task '%s' on %s", name, dependencies[name])
-                self.get_task(name).dependencies.extend(self._dependencies_pending_tasks[name])
+                dependency_tasks = self._dependencies_pending_tasks[name]
+                self.logger.debug("Resolved late dependency of task '%s' on %s", name, dependency_tasks)
+                self.get_task(name).dependencies.extend(dependency_tasks)
                 del self._dependencies_pending_tasks[name]
 
     def execute_initializers(self, environments=None, **kwargs):
@@ -418,7 +418,7 @@ class ExecutionManager(object):
         visited.add(task)
         dependencies = [dependency for dependency in self._task_dependencies[task.name]]
         for dependency in dependencies:
-            self._collect_transitive_tasks(dependency.task, visited)
+            self._collect_transitive_tasks(self.get_task(dependency.name), visited)
         return visited
 
     def collect_all_transitive_tasks(self, task_names):
@@ -519,7 +519,7 @@ class ExecutionManager(object):
                     if existing_dependency.name == d.name:
                         if existing_dependency.optional != d.optional:
                             if existing_dependency.optional:
-                                task_dependencies[index] = TaskDependency(existing_dependency.name)
+                                existing_dependency.required()
                                 self.logger.debug("Converting optional dependency '%s' of task '%s' into required",
                                                   existing_dependency, task.name)
                             else:

--- a/src/unittest/python/execution_tests.py
+++ b/src/unittest/python/execution_tests.py
@@ -594,10 +594,13 @@ class ExecutionManagerResolveDependenciesTest(ExecutionManagerTestBase):
         one = Mock(name="one", dependencies=[])
         two = Mock(name="two", dependencies=[TaskDependency("one")])
         three = Mock(name="three", dependencies=[TaskDependency("one")])
+        four = Mock(name="four", dependencies=[])
 
         self.execution_manager.register_task(one, two)
         self.execution_manager.register_task(three)
-        self.execution_manager.register_late_task_dependencies({"two": [TaskDependency("three")]})
+        self.execution_manager.register_late_task_dependencies({"two": [TaskDependency("three")],
+                                                                "four": [TaskDependency("two")]})
+        self.execution_manager.register_task(four)
         self.execution_manager.resolve_dependencies()
 
         self.assertEqual([], self.execution_manager._task_dependencies.get("one"))
@@ -630,8 +633,7 @@ class ExecutionManagerResolveDependenciesTest(ExecutionManagerTestBase):
 
         self.execution_manager.register_task(one, two)
         self.execution_manager.register_task(three)
-        self.execution_manager.register_late_task_dependencies(
-            {"four": [TaskDependency("three")]})
+        self.execution_manager.register_late_task_dependencies({"four": [TaskDependency("three")]})
         self.assertRaises(NoSuchTaskException, self.execution_manager.resolve_dependencies)
 
 

